### PR TITLE
Externalize node admission controller

### PR DESCRIPTION
**Pull request by yue9944882**
**Release note**:

```release-note
NONE
```

**What this PR does / why we need it**:

/sig api-machinery

under unbrella: #66680

We start with injecting external informer/lister into node admission controller for that it's slightly simpler than the others.

Here is the thing I actually did in this pull:

1. Enable plugin initializer to inject external/internal clientset/informers into the admission controllers.
2. Receive external informer in node admission controller.
3. Create a `AdmissionConfig` following how we did in `AuthenticationConfig` and `AuthorizationConfig` and creates the actual plugin initializers and the admission post-start hook in its `New()` method

### CODE REVIEW

**Comment by liggitt** [comment]
Pod lister/informers are among the highest quantity/event volume informers. We should ensure we don't end up with both internal and external informers active in the apiserver for them. 

@kubernetes/sig-scalability-pr-reviews

**Comment by yue9944882** [comment]
> We should ensure we don't end up with both internal and external informers active in the apiserver for them.

Is that a problem if both internal/external exists at the same time? I thought it'll be smooth mostly except for some utility functions are defined to receive only internal types, and we might need some extra effort to sth these admission controller also relying on. The switching might be quick for most for these admisson controllers.

**Comment by yue9944882** [comment]
> It's not just admission plugins, I know at least the node authorizer also watches pods via an informer. Would need to sweep for other pod informer uses in the process

i'd happy to do so. but i'm afraid it will involve too many changes together, making this pull hard to review. i suppose node admission controller might be a low hanging fruit compare with the others.

**Comment by deads2k** [comment]
> Pod lister/informers are among the highest quantity/event volume informers. We should ensure we don't end up with both internal and external informers active in the apiserver for them.

@liggitt I only see the one reference in the node restriction admission plugin.

/hold cancel

**Comment by deads2k** [comment]
> Pod lister/informers are among the highest quantity/event volume informers. We should ensure we don't end up with both internal and external informers active in the apiserver for them.
> 
> @liggitt I only see the one reference in the node restriction admission plugin.
> 
> /hold cancel

Tricked.  Same feature two spots.  @yue9944882 pick up the related authorizer as well and this can go in.

File: pkg/kubeapiserver/admission/config.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	externalinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	aggregatorapiserver "k8s.io/kube-aggregator/pkg/apiserver"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	internalinformers "k8s.io/kubernetes/pkg/client/informers/informers_generated/internalversion"
+	quotainstall "k8s.io/kubernetes/pkg/quota/install"
+)
+
+type AdmissionConfig struct {
```
Comment (deads2k) [inline pkg/kubeapiserver/admission/config.go]: This seems like a topic for another pull.  Can you pull the plugin initializer changes out of this pull?

File: pkg/kubeapiserver/admission/initializer.go [hunk 2]
```diff
@@ -60,7 +59,6 @@ type WantsQuotaConfiguration interface {
 // PluginInitializer is used for initialization of the Kubernetes specific admission plugins.
 type PluginInitializer struct {
 	internalClient                    internalclientset.Interface
-	externalClient                    clientset.Interface
```
Comment (yue9944882) [inline pkg/kubeapiserver/admission/initializer.go]: @deads2k btw `externalClient` is never used here. cleaning up.

File: pkg/kubeapiserver/authorizer/config.go [hunk 3]
```diff
@@ -75,7 +75,7 @@ func (config AuthorizationConfig) New() (authorizer.Authorizer, authorizer.RuleR
 			node.AddGraphEventHandlers(
 				graph,
 				config.InformerFactory.Core().InternalVersion().Nodes(),
-				config.InformerFactory.Core().InternalVersion().Pods(),
+				config.VersionedInformerFactory.Core().V1().Pods(),
```
Comment (yue9944882) [inline pkg/kubeapiserver/authorizer/config.go]: we should only externalize pod informer here, am i right?

File: pkg/kubeapiserver/admission/config.go (hunk 1 above)
Comment (yue9944882) [reply pkg/kubeapiserver/admission/config.go]: @deads2k done. can't help refactoring here to shrink the long param/return list. 😅

File: pkg/kubeapiserver/authorizer/config.go (hunk 3 above)
Comment (liggitt) [reply pkg/kubeapiserver/authorizer/config.go]: Moving one resource type at a time would minimize the size of the PR

Comment (yue9944882) [reply pkg/kubeapiserver/authorizer/config.go]: > Moving one resource type at a time would minimize the size of the PR

make sense 👍

**Comment by liggitt** [comment]
Does quota admission already use an external pod informer?

**Comment by deads2k** [comment]
> Does quota admission already use an external pod informer?

False predicate.  Quota admission doesn't use any pod informer.

**Comment by deads2k** [comment]
nice job.

/lgtm

**Pull request merged**